api/dbadapters/presto.ts

```diff
@@ -0,0 +1,228 @@
+    const validationQueriesWithoutWrappers = collectEvaluationQueries(
+      queryOrAction,
+      useSingleQueryPerAction
+    );
+
+    const queryEvaluations = new Array<dataform.IQueryEvaluation>();
+    for (const { index, validationQuery } of validationQueries) {
+      let evaluationResponse: dataform.IQueryEvaluation = {
+        status: dataform.QueryEvaluation.QueryEvaluationStatus.SUCCESS
+      };
+      try {
+        await prestoExecute(this.client, validationQuery.query);
+      } catch (e) {
+        evaluationResponse = {
+          // TODO: Parse the error with line number.
+          status: dataform.QueryEvaluation.QueryEvaluationStatus.FAILURE,
+          error: e
+        };
+      }
+      queryEvaluations.push(
+        dataform.QueryEvaluation.create({
+          ...evaluationResponse,
+          incremental: validationQuery.incremental,
+          query: validationQueriesWithoutWrappers[index].query
+        })
+      );
+    }
+    return queryEvaluations;
+  }
+
+  public async catalogs(): Promise<string[]> {
+    const result = await prestoExecute(this.client, "show catalogs");
+    const catalogs = flatten(result.data);
+    return catalogs;
+  }
+
+  public async schemas(): Promise<string[]> {
+    let schemas: string[] = [];
+    await Promise.all(
+      (await this.catalogs()).map(async catalog => {
+        const result = await prestoExecute(this.client, `show schemas from ${catalog}`);
+        schemas = schemas.concat(flatten(result.data).map(schema => `${catalog}.${schema}`));
+      })
+    );
+    // TODO: Currently this returns `catalog.schema` rather than just schemas in order to
+    // conform to string return type. Should this be changed? In tables() it is converted to targets.
+    return schemas;
+  }
+
+  public async createSchema(catalog: string, schema: string): Promise<void> {
+    await prestoExecute(this.client, `create schema if not exists ${catalog}.${schema}`);
+  }
+
+  // TODO: This should take parameters to allow for retrieving from a specific catalog/schema.
+  public async tables(): Promise<dataform.ITarget[]> {
+    let tables: dataform.ITarget[] = [];
+    await Promise.all(
+      (await this.schemas()).map(async schemaString => {
+        const [catalog, schema] = schemaString.split(".");
+        const result = await prestoExecute(this.client, `show tables from ${catalog}.${schema}`);
+        tables = tables.concat(
+          flatten(result.data).map(table => ({ database: catalog, schema, table }))
+        );
+      })
+    );
+    return tables;
+  }
+
+  public async table(target: dataform.ITarget): Promise<dataform.ITableMetadata> {
+    let columnsResult: IPrestoExecutionResult;
+    try {
+      columnsResult = await prestoExecute(this.client, `describe ${resolveTarget(target)}`);
+    } catch (e) {
+      // This probably failed because the table doesn't exist, so ignore as the information isn't necessary.
```

Comment:
this 'probably' is a little scary, if it fails due to some other error (but the table does exist), we can easily end up e.g. corrupting incremental table state! can we do some string matching on the error at least?

Reply:
I'm not sure what my thinking was here any more, I think this should throw in all cases

Reply:
Oh I see now, when we run a project we call describe on the table in case there's useful metadata. Have added a check for the exact error code of the client